apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.RequestTemplate.cs

```diff
@@ -0,0 +1,226 @@
+            /// or any header starting with "x-goog-".
+            /// </para>
+            /// <para>
+            /// Note that if the entity is encrypted with customer-supplied encryption keys (see
+            /// https://cloud.google.com/storage/docs/encryption for more information), the <b>x-goog-encryption-algorithm</b>,
+            /// <b>x-goog-encryption-key</b>, and <b>x-goog-encryption-key-sha256</b> headers will be required when making the
+            /// request. However, only the x-goog-encryption-algorithm header is included in the signature for the signed URL.
+            /// So the <see cref="RequestHeaders"/> specified only need to have the x-goog-encryption-algorithm header.
+            /// </para>
+            /// </remarks>
+            public IReadOnlyDictionary<string, IReadOnlyCollection<string>> RequestHeaders { get; }
+
+            /// <summary>
+            /// The headers for the content which will be included with the request. Won't be null.
+            /// If null is specified in <see cref="WithContentHeaders(IEnumerable{KeyValuePair{string, IEnumerable{string}}})"/>
+            /// or <see cref="WithContentHeaders(IReadOnlyDictionary{string, IReadOnlyCollection{string}})"/> then an empty
+            /// dictionary will be used.
+            /// </summary>
+            /// <remarks>
+            /// <para>
+            /// When either of <see cref="RequestHeaders"/> or <see cref="ContentHeaders"/> are specified, there are certain 
+            /// headers which will be included in the signed URL's signature, and therefore must be included in requests made 
+            /// with the created URL. These are the Content-MD5 and Content-Type content headers as well as any content or 
+            /// request header with a name starting with "x-goog-".
+            /// </para>
+            /// <para>
+            /// If both <see cref="RequestHeaders"/> or <see cref="ContentHeaders"/> are null or empty, no headers are included 
+            /// in the signed URL's signature, so any requests made with the created URL must not contain Content-MD5, Content-Type,
+            /// or any header starting with "x-goog-".
+            /// </para>
+            /// </remarks>
+            public IReadOnlyDictionary<string, IReadOnlyCollection<string>> ContentHeaders { get; }
+
+            private RequestTemplate(
+                string bucket,
+                string objectName,
+                HttpMethod httpMethod,
+                IReadOnlyDictionary<string, IReadOnlyCollection<string>> requestHeaders,
+                IReadOnlyDictionary<string, IReadOnlyCollection<string>> contentHeaders)
+            {
+                Bucket = StorageClientImpl.ValidateBucketName(bucket);
+                ObjectName = objectName;
+                HttpMethod = httpMethod ?? HttpMethod.Get;
+                RequestHeaders = requestHeaders ?? UrlV4SignerHelperExtensions.s_empty;
```

Comment:
It would be nice to be able to avoid referring to a static field of another class. Maybe have one within UrlSigner itself? That would at least feel a little better.

Reply:
Done.